specification/search/data-plane/Azure.Search/indexreadme.md

```diff
@@ -0,0 +1,479 @@
+
+> `autorest`
+
+To see additional help and options, run:
+
+> `autorest --help`
+---
+
+## Configuration
+
+### Basic Information
+These are the global settings for SearchIndexClient.
+
+``` yaml
+openapi-type: data-plane
+tag: package-2019-05
+```
+
+
+### Tag: package-2019-05-searchservice-preview
+
+These settings apply only when `--tag=package-2019-05-searchservice-preview` is specified on the command line.
+
+``` yaml $(tag) == 'package-2019-05-searchservice-preview'
+input-file:
+- preview/2019-05-06-preview/searchservice.json
+```
+
+### Tag: package-2019-05-searchindex-preview
+
+These settings apply only when `--tag=package-2019-05-searchindex-preview` is specified on the command line.
+
+``` yaml $(tag) == 'package-2019-05-searchindex-preview'
+input-file:
+- preview/2019-05-06-preview/searchindex.json
+```
+
+### Tag: package-2019-05-searchservice
+
+These settings apply only when `--tag=package-2019-05-searchservice` is specified on the command line.
+
+``` yaml $(tag) == 'package-2019-05-searchservice'
+input-file:
+- stable/2019-05-06/searchservice.json
+```
+
+### Tag: package-2019-05-searchindex
+
+These settings apply only when `--tag=package-2019-05-searchindex` is specified on the command line.
+
+``` yaml $(tag) == 'package-2019-05-searchindex'
+opt-in-extensible-enums: true
+input-file:
+- track1/preview/2019-05-06-preview/searchindex.json
+```
+
+---
+# Code Generation
+
+!!! READ THIS !!!
+This swagger is ready for C# and Java
+!!! READ THIS !!!
+
+## Java
+
+These settings apply only when `--java` is specified on the command line.
+Please also specify `--java-sdks-folder=<path to "SDKs" directory of your azure-sdk-for-java clone>`.
+
+``` yaml $(java)
+title: SearchIndexRestClient
+output-folder: $(java-sdks-folder)/search
+namespace: com.azure.search
+java: true
+sync-methods: none
+add-context-parameter: true
+generate-client-interfaces: false
+license-header: |-
+  Copyright (c) Microsoft Corporation. All rights reserved.
+  Licensed under the MIT License.
+  Code generated by Microsoft (R) AutoRest Code Generator.
+  Changes may cause incorrect behavior and will be lost if the code is regenerated.
+
+require: ./autorest-custom-directives.md
+
+vararg-properties: >-
+  AutocompleteOptions.searchFields,
+  SearchOptions.facets, SearchOptions.highlightFields, SearchOptions.orderBy, SearchOptions.scoringParameters, SearchOptions.searchFields, SearchOptions.select,
+  SuggestOptions.orderBy, SuggestOptions.searchFields, SuggestOptions.select
+```
+
+### Tweak generated code
+
+Use the below directives sparingly. Every modification we make in here, we potentially have to replicate for every target language.
+
+``` yaml $(java)
+directive:
+    # Rename IndexBatch to IndexBatchBase when processing the API spec
+    - rename-model:
+          from: IndexBatch
+          to: IndexBatchBase
+
+    # Use Document rather than Map<String, Object>
```

Comment:
Could we keep this as `Map<String, Object>` in the codegen layer given `SearchDocument`/`Document` (depending on where we are in renaming) can just take the object in its constructor. I'm not a fan of requiring the generated code requiring hand crafted classes since that seems a bit backwards.

Reply:
This is purely an implementation suggestion and won't affect the public contract of `SearchDocument`, right?

Reply:
Correct, I'll make a deeper review of where this codegen `Map<String, Object>` is being replaced and see if it could be used instead of requiring a hand-written class to replace it. Another potential solution would be adding a Swagger transform that generates the `SearchDocument` class for us, which will always ensure that it exists.

Reply:
Just make sure that `SearchDocument` continues to have the constructor that takes a `Map<String, Object>` since that was added for interop purposes based on feedback from customers using Clojure.

Reply:
Will test in my code if it can be removed. Thanks!